src/stdune/io.ml

```diff
@@ -109,9 +109,22 @@ struct
           };
         f lb)
 
-  let read_all ic =
-    let len = in_channel_length ic in
-    really_input_string ic len
+  (* This function is a copy&paste from Jane Street's [Stdio.In_channel.input_all]. *)
+  let read_all t =
+    (* We use 65536 because that is the size of OCaml's IO buffers. *)
+    let buf_size = 65536 in
+    let buf = Bytes.create buf_size in
+    let buffer = Buffer.create buf_size in
+    let rec loop () =
```

Comment:
In `copy_channels`, we use a static buffer. It feels like `read_all` should be consistent with this choice. In particular, we might as well use the same static buffer for both functions.

BTW, it seems useful to me to keep the optimisation for regular files given that most of the files dune read are small regular files. i.e. only fallback to this implementation if the `in_channel_length` fails.

Reply:
I guess `in_channel_length` returns zero because we ended up with an empty string in the case of non-file. I'm not very familiar with the in_channel API. Is there a way to fstat the thing?

Reply:
Damn, if it's not supported why can't the call just fail :roll_eyes: I didn't realise the call was returning `0`, I thought it was raising. We can extract a fd with `Unix.file_descr_of_in_channel`, but that's not going to be very useful; I check that the kind is `S_REG`...

BTW, I just remembered about `Buffer.add_channel`. We can do `Buffer.add_channel ic max_int`. This avoids an extra memory copy  and the code is simpler.

Reply:
So actually not `max_int` because it will try to allocate a huge buffer, but we can call `add_channel ic 64k` until it raises.

Reply:
Two differences compared to `copy_channels`:
- the size of the buffer here is unbounded, so we'd need to worry about a shared buffer getting large and staying large
- we need to allocate the final string anyway, so might as well allocate the intermediate one

I think it's better if we don't use a static buffer for this one

Reply:
👏

Reply:
Seems good. I reworked a bit the code to preserve the optimisation when possible; the idea is to use the old code path and then do `input_char` to make sure we reached the end of the buffer. If we didn't, we go through the generic code path.